File: bot/engine.py
"""Pure conversation logic. No API calls, randomness or model-generated claims."""
import copy
import json
import math
import re
from pathlib import Path

WEB = Path(__file__).resolve().parent.parent / 'web'
DATA = json.loads((WEB / 'schemes.json').read_text(encoding='utf-8'))
UI = json.loads((WEB / 'ui.json').read_text(encoding='utf-8'))
SCHEMES = {s['id']: s for s in DATA['schemes']}


def normalise(value):
    return str(value).strip().lower().translate(str.maketrans('०१२३४५६७८९', '0123456789'))


def initial(lang='en'):
    return {'lang': lang, 'profile': {}, 'scheme': None, 'pending': None, 'options': []}


def evaluate(scheme, profile):
    missing = [f for f in scheme['fields'] if f not in profile or profile[f] == 'unknown']
    failures = []
    for r in scheme['rules']:
        v = profile.get(r['field'])
        if v is None or v == 'unknown':
            continue
        if r['op'] == 'marks':
            if profile.get('sc_st') in (None, 'unknown'):
                continue
            passed = v >= (50 if profile['sc_st'] == 'yes' else r['value'])
        elif r['op'] == 'eq':
            passed = v == r['value']
        elif r['op'] == 'lte':
            passed = v <= r['value']
        elif r['op'] == 'gte':
            passed = v >= r['value']
        else:
            raise ValueError('Unsupported rule operator')
        if not passed:
            failures.append(r['reason'])
    return ('unlikely' if failures else 'review' if missing else 'possible'), failures


def parse(field, raw):
    value = normalise(raw)
    if value in ('unknown', 'not sure', "don't know", 'पता नहीं', 'नहीं पता'):
        return 'unknown'
    if field['type'] == 'choice':
        for i, option in enumerate(field['options'], 1):
            aliases = [normalise(v) for v in option] + [str(i)]
            if option[0] == 'yes':
                aliases += ['haan', 'han', 'ha', 'हां']
            if option[0] == 'no':
                aliases += ['nahi', 'nahin']
            if value in aliases:
                return option[0]
    elif field['type'] == 'number':
        value = value.replace(',', '').replace('₹', '').strip().removesuffix('%').strip()
        if not re.fullmatch(r'\d+(?:\.\d+)?', value):
            return None
        n = float(value)
        if math.isfinite(n) and field['min'] <= n <= field['max'] and (not field.get('integer') or n.is_integer()):
            return int(n) if n.is_integer() else n
    elif field['type'] == 'state':
        if value in DATA['state_aliases']:
            return DATA['state_aliases'][value]
        return next((s for s in DATA['states'] if s.lower() == value), None)
    return None


def respond(saved, text):
    s = copy.deepcopy(saved or initial('hi' if re.search('[\u0900-\u097f]', text) else 'en'))
    raw = normalise(text)
    lang = s['lang']

    def t(key):
        return UI[key][s['lang']]

    def result(kind, body, options=None, links=None, **extra):
        opts = options or []
        s['options'] = opts
        return s, {'kind': kind, 'text': body, 'options': opts, 'links': links or [], 'lang': s['lang'], **extra}

    def link(label, url):
        return {'label': label, 'url': url}

    def scheme_options():
        return [{'label': x['icon'] + ' ' + x['name'][s['lang']], 'value': x['id']} for x in DATA['schemes']]

    def menu(prefix=''):
        s['scheme'] = s['pending'] = None
        return result('menu', (prefix + '\n\n' if prefix else '') + t('welcome') + '\n\n' + t('saved'), scheme_options())

    def advance():
        scheme = SCHEMES[s['scheme']]
        for key in scheme['fields']:
            if key not in s['profile']:
                s['pending'] = key
                field = DATA['fields'][key]
                opts = [{'value': o[0], 'label': o[1 if s['lang'] == 'en' else 2]} for o in field.get('options', [])]
                return result('question', field['question'][s['lang']], opts,
                              field=key, scheme=scheme['id'], progress={'done': sum(f in s['profile'] for f in scheme['fields']), 'total': len(scheme['fields'])})
        return assessment()

    def assessment():
        scheme = SCHEMES[s['scheme']]
        status, reasons = evaluate(scheme, s['profile'])
        s['pending'] = None
        lines = [scheme['name'][s['lang']], t(status)]
        lines += [r[s['lang']] for r in reasons]
        if status == 'review':
            lines.append(t('unknown'))
        lines += [scheme['benefit'][s['lang']], scheme['scope'][s['lang']], scheme['documents'][s['lang']], t('disclaimer')]
        return result('assessment', '\n\n'.join(lines),
                      [{'label': t('menu'), 'value': '/menu'}, {'label': t('edit'), 'value': '/edit'}],
                      [link(t('official'), scheme['apply']), link(t('source'), scheme['source']), link(t('csc'), DATA['csc_url'])],
                      status=status, scheme=scheme['id'])

    if raw in ('/delete', '/reset', 'delete'):
        s = initial(lang)
        return menu(t('deleted'))
    if raw in ('/privacy', 'privacy'):
        return result('info', t('privacy'), [{'label': t('menu'), 'value': '/menu'}])
    if raw in ('/voice',):
        return result('info', t('voice'), [{'label': t('menu'), 'value': '/menu'}])
    if raw in ('/help', 'help', '/csc', 'csc', 'मदद', 'सहायता'):
        return result('help', t('help'), [{'label': t('menu'), 'value': '/menu'}], [link(t('csc'), DATA['csc_url']), link('myScheme', 'https://www.myscheme.gov.in/')])
    if raw in ('/language', '/hi', '/en', 'hindi', 'हिंदी', 'english'):
        s['lang'] = ('hi' if s['lang'] == 'en' else 'en') if raw == '/language' else ('hi' if raw in ('/hi', 'hindi', 'हिंदी') else 'en')
        return advance() if s['scheme'] and s['pending'] else menu()
    if raw in ('/edit', 'edit'):
        s['profile'] = {}
        return advance() if s['scheme'] else menu()
    if raw in ('/back', 'back') and s['scheme']:
        fields = SCHEMES[s['scheme']]['fields']
        index = fields.index(s['pending']) if s['pending'] else len(fields)
        if index:
            s['profile'].pop(fields[index - 1], None)
        return advance()
    if raw in ('/start', '/menu', 'menu', 'hi', 'hello', 'namaste', 'नमस्ते', ''):
        return menu()

    # Button payloads are unambiguous. Bare numbers only select a menu when no question is active.
    selected = raw if raw in SCHEMES else None
    if not selected and not s['pending']:
        if raw in ('1', '2', '3') and s['options'] == scheme_options():
            selected = DATA['schemes'][int(raw) - 1]['id']
        else:
            matches = [x['id'] for x in DATA['schemes'] if any(k in raw for k in x['keywords'])]
            if len(matches) == 1:
                selected = matches[0]
    if selected:
        s['scheme'] = selected
        return advance()
    if s['pending']:
        field = DATA['fields'][s['pending']]
        value = parse(field, text)
        if value is None:
            state, reply = advance()
            reply['text'] = t('invalid') + '\n\n' + reply['text']
            return state, reply
        s['profile'][s['pending']] = value
        return assessment() if value == 'unknown' else advance()
    return result('fallback', t('fallback'), scheme_options(), [link(t('csc'), DATA['csc_url']), link('myScheme', 'https://www.myscheme.gov.in/')])

File: web/engine.mjs
// The browser and Python adapters read the same versioned rules and translations.
export const normalise = v => String(v).trim().toLowerCase().replace(/[०-९]/g, c => String(c.charCodeAt(0) - 2406));
export const initial = (lang = 'en') => ({lang, profile: {}, scheme: null, pending: null, options: []});

export function evaluate(scheme, profile) {
  const missing = scheme.fields.filter(f => !(f in profile) || profile[f] === 'unknown');
  const reasons = [];
  for (const r of scheme.rules) {
    const v = profile[r.field];
    if (v === undefined || v === 'unknown') continue;
    let passed;
    if (r.op === 'marks') {
      if (!profile.sc_st || profile.sc_st === 'unknown') continue;
      passed = v >= (profile.sc_st === 'yes' ? 50 : r.value);
    } else if (r.op === 'eq') passed = v === r.value;
    else if (r.op === 'lte') passed = v <= r.value;
    else if (r.op === 'gte') passed = v >= r.value;
    else throw new Error('Unsupported rule operator');
    if (!passed) reasons.push(r.reason);
  }
  return {status: reasons.length ? 'unlikely' : missing.length ? 'review' : 'possible', reasons};
}

export function parse(field, raw, data) {
  let value = normalise(raw);
  if (['unknown', 'not sure', "don't know", 'पता नहीं', 'नहीं पता'].includes(value)) return 'unknown';
  if (field.type === 'choice') {
    for (const [i, o] of field.options.entries()) {
      const aliases = [...o.map(normalise), String(i + 1)];
      if (o[0] === 'yes') aliases.push('haan', 'han', 'ha', 'हां');
      if (o[0] === 'no') aliases.push('nahi', 'nahin');
      if (aliases.includes(value)) return o[0];
    }
  } else if (field.type === 'number') {
    value = value.replaceAll(',', '').replaceAll('₹', '').trim().replace(/%$/, '').trim();
    if (!/^\d+(?:\.\d+)?$/.test(value)) return null;
    const n = Number(value);
    if (Number.isFinite(n) && n >= field.min && n <= field.max && (!field.integer || Number.isInteger(n))) return n;
  } else if (field.type === 'state') {
    return (Object.hasOwn(data.state_aliases, value) ? data.state_aliases[value] : null) || data.states.find(s => s.toLowerCase() === value) || null;
  }
  return null;
}

export function respond(saved, text, data, ui) {
  let s = structuredClone(saved || initial(/[\u0900-\u097f]/.test(text) ? 'hi' : 'en'));
  const raw = normalise(text), oldLang = s.lang;
  const schemes = Object.fromEntries(data.schemes.map(x => [x.id, x]));
  const t = key => ui[key][s.lang];
  const link = (label, url) => ({label, url});
  const schemeOptions = () => data.schemes.map(x => ({label: x.icon + ' ' + x.name[s.lang], value: x.id}));
  const result = (kind, text, options = [], links = [], extra = {}) => {
    s.options = options;
    return {state: s, reply: {kind, text, options, links, lang: s.lang, ...extra}};
  };
  const menu = (prefix = '') => {
    s.scheme = s.pending = null;
    return result('menu', (prefix ? prefix + '\n\n' : '') + t('welcome') + '\n\n' + t('saved'), schemeOptions());
  };
  const assessment = () => {
    const scheme = schemes[s.scheme], {status, reasons} = evaluate(scheme, s.profile);
    s.pending = null;
    const lines = [scheme.name[s.lang], t(status), ...reasons.map(r => r[s.lang])];
    if (status === 'review') lines.push(t('unknown'));
    lines.push(scheme.benefit[s.lang], scheme.scope[s.lang], scheme.documents[s.lang], t('disclaimer'));
    return result('assessment', lines.join('\n\n'), [{label: t('menu'), value: '/menu'}, {label: t('edit'), value: '/edit'}],
      [link(t('official'), scheme.apply), link(t('source'), scheme.source), link(t('csc'), data.csc_url)], {status, scheme: scheme.id});
  };
  const advance = () => {
    const scheme = schemes[s.scheme];
    for (const key of scheme.fields) {
      if (!(key in s.profile)) {
        s.pending = key;
        const field = data.fields[key];
        return result('question', field.question[s.lang], (field.options || []).map(o => ({value:o[0], label:o[s.lang === 'en' ? 1 : 2]})), [],
          {field:key, scheme:scheme.id, progress:{done:scheme.fields.filter(f => f in s.profile).length, total:scheme.fields.length}});
      }
    }
    return assessment();
  };
  if (['/delete','/reset','delete'].includes(raw)) { s = initial(oldLang); return menu(t('deleted')); }
  if (['/privacy','privacy','/voice'].includes(raw)) return result('info', t(raw === '/voice' ? 'voice' : 'privacy'), [{label:t('menu'),value:'/menu'}]);
  if (['/help','help','/csc','csc','मदद','सहायता'].includes(raw)) return result('help', t('help'), [{label:t('menu'),value:'/menu'}], [link(t('csc'),data.csc_url),link('myScheme','https://www.myscheme.gov.in/')]);
  if (['/language','/hi','/en','hindi','हिंदी','english'].includes(raw)) {
    s.lang = raw === '/language' ? (s.lang === 'en' ? 'hi' : 'en') : (['/hi','hindi','हिंदी'].includes(raw) ? 'hi' : 'en');
    return s.scheme && s.pending ? advance() : menu();
  }
  if (['/edit','edit'].includes(raw)) {s.profile = {}; return s.scheme ? advance() : menu();}
  if (['/back','back'].includes(raw) && s.scheme) {
    const fields = schemes[s.scheme].fields, i = s.pending ? fields.indexOf(s.pending) : fields.length;
    if (i) delete s.profile[fields[i-1]];
    return advance();
  }
  if (['/start','/menu','menu','hi','hello','namaste','नमस्ते',''].includes(raw)) return menu();
  let selected = Object.hasOwn(schemes, raw) ? raw : null;
  if (!selected && !s.pending) {
    if (['1','2','3'].includes(raw) && JSON.stringify(s.options) === JSON.stringify(schemeOptions())) selected = data.schemes[Number(raw)-1].id;
    else {
      const matches = data.schemes.filter(x => x.keywords.some(k => raw.includes(k)));
      if (matches.length === 1) selected = matches[0].id;
    }
  }
  if (selected) {s.scheme = selected; return advance();}
  if (s.pending) {
    const value = parse(data.fields[s.pending], text, data);
    if (value === null) {const answer = advance(); answer.reply.text = t('invalid') + '\n\n' + answer.reply.text; return answer;}
    s.profile[s.pending] = value;
    return value === 'unknown' ? assessment() : advance();
  }
  return result('fallback', t('fallback'), schemeOptions(), [link(t('csc'),data.csc_url),link('myScheme','https://www.myscheme.gov.in/')]);
}
